hcloud/action.go

```diff
@@ -78,14 +79,70 @@ func (c *ActionClient) GetByID(ctx context.Context, id int) (*Action, *Response,
 	return ActionFromSchema(body.Action), resp, nil
 }
 
+// ActionPage serves as accessor of the actions API pagination.
+type ActionPage struct {
+	Page
+	content []*Action
+}
+
+// Content contains the content of the current page.
+func (p *ActionPage) Content() []*Action {
+	return p.content
+}
+
 // ActionListOpts specifies options for listing actions.
 type ActionListOpts struct {
 	ListOpts
+	SortOpts
+	Status     []ActionStatus
+	Server     *Server
+	FloatingIP *FloatingIP
+}
+
+// URLValues returns the list opts as url.Values.
+func (o ActionListOpts) URLValues() url.Values {
+	vals := o.ListOpts.URLValues()
+	vals["sort"] = o.SortOpts.URLValues()["sort"]
+	for _, s := range o.Status {
+		vals.Add("status", string(s))
+	}
+	return vals
+}
+
+// List returns an accessor to control the actions API pagination.
+func (c *ActionClient) List(ctx context.Context, opts ActionListOpts) *ActionPage {
+	page := &ActionPage{}
+	page.pageGetter = pageGetter(func(start, end int) (resp *Response, exhausted bool, err error) {
+		allActions := []*Action{}
+		if opts.PerPage == 0 {
+			opts.PerPage = 50
+		}
+
+		resp, exhausted, err = c.client.all(func(page int) (*Response, error) {
+			opts.Page = page
+			actions, resp, err := c.list(ctx, opts)
+			if err != nil {
+				return resp, err
+			}
+			allActions = append(allActions, actions...)
+			return resp, nil
+		}, start, end)
+		page.content = allActions
+		return
+	})
+	return page
 }
 
-// List returns a list of actions for a specific page.
-func (c *ActionClient) List(ctx context.Context, opts ActionListOpts) ([]*Action, *Response, error) {
-	path := "/actions?" + valuesForListOpts(opts.ListOpts).Encode()
+// list returns a list of actions for a specific page.
+func (c *ActionClient) list(ctx context.Context, opts ActionListOpts) ([]*Action, *Response, error) {
+	path := "/actions?"
+	if opts.Server != nil {
+		path = fmt.Sprintf("/servers/%d/actions?", opts.Server.ID)
+	}
+	if opts.FloatingIP != nil {
+		path = fmt.Sprintf("/floating_ips/%d/actions?", opts.FloatingIP.ID)
+	}
```

Comment:
I’m not sure if having these filters is a good idea – at least how the work now.

When I specify a sever and a Floating IP, I would expect to get actions concerting both server *and* Floating IP, but right now the Floating IP filter would win and discard the server filter.

I would remove these filters for now.